examples/jsm/objects/BatchedMesh.js

```diff
@@ -0,0 +1,461 @@
+
+				const dstArray = new array.constructor( this._maxVertexCount * itemSize );
+				const dstAttribute = new srcAttribute.constructor( dstArray, itemSize, normalized );
+				dstAttribute.setUsage( srcAttribute.usage );
+
+				this.geometry.setAttribute( attributeName, dstAttribute );
+
+			}
+
+			if ( geometry.getIndex() !== null ) {
+
+				const indexArray = this._maxVertexCount > 65536
+					? new Uint32Array( this._maxIndexCount )
+					: new Uint16Array( this._maxIndexCount );
+
+				this.geometry.setIndex( new BufferAttribute( indexArray, 1 ) );
+
+			}
+
+			const idArray = this._maxGeometryCount > 65536
+				? new Uint32Array( this._maxVertexCount )
+				: new Uint16Array( this._maxVertexCount );
+			// @TODO: What if attribute name 'id' is already used?
+			this.geometry.setAttribute( 'id', new BufferAttribute( idArray, 1 ) );
+
+			this._geometryInitialized = true;
+
+		} else {
+
+			// @TODO: Check if geometry has the same attributes set
+
+		}
+
+		const hasIndex = this.geometry.getIndex() !== null;
+		const dstIndex = this.geometry.getIndex();
+		const srcIndex = geometry.getIndex();
+
+		// Assuming geometry has position attribute
+		const srcPositionAttribute = geometry.getAttribute( 'position' );
+
+		this._vertexStarts.push( this._currentVertexCount );
+		this._vertexCounts.push( srcPositionAttribute.count );
+
+		if ( hasIndex ) {
+
+			this._indexStarts.push( this._currentIndexCount );
+			this._indexCounts.push( srcIndex.count );
+
+		}
+
+		this._visibilities.push( true );
+		this._alives.push( true );
+
+		// @TODO: Error handling if exceeding maxVertexCount or maxIndexCount
+
+		for ( const attributeName in geometry.attributes ) {
+
+			const srcAttribute = geometry.getAttribute( attributeName );
+			const dstAttribute = this.geometry.getAttribute( attributeName );
+
+			dstAttribute.array.set( srcAttribute.array, this._currentVertexCount * dstAttribute.itemSize );
+			dstAttribute.needsUpdate = true;
+
+		}
+
+		if ( hasIndex ) {
+
+			for ( let i = 0; i < srcIndex.count; i ++ ) {
+
+				dstIndex.setX( this._currentIndexCount + i, this._currentVertexCount + srcIndex.getX( i ) );
+
+			}
+
+			this._currentIndexCount += srcIndex.count;
+			dstIndex.needsUpdate = true;
+
+		}
+
+		const geometryId = this._currentGeometryCount;
+		this._currentGeometryCount ++;
+
+		const idAttribute = this.geometry.getAttribute( 'id' );
+
+		for ( let i = 0; i < srcPositionAttribute.count; i ++ ) {
+
+			idAttribute.setX( this._currentVertexCount + i, geometryId );
+
+		}
+
+		idAttribute.needsUpdate = true;
+
+		this._currentVertexCount += srcPositionAttribute.count;
+
+		this._matrices.push( new Matrix4() );
+		_identityMatrix.toArray( this._matricesArray, geometryId * 16 );
+		this._matricesTexture.needsUpdate = true;
+
+		return geometryId;
+
+	}
+
+	discardGeometry( geometryId ) {
+
+		if ( geometryId >= this._alives.length || this._alives[ geometryId ] === false ) {
+
+			// @TODO: Warning?
```

Comment:
In general — I think it's OK (and possibly better) for these methods to hit errors on invalid input; we don't typically sanitize it, and I silently doing nothing may cause developers some headaches.